### ISSUE TITLE
No DDS sRGB Color Space Support (Export)

### ISSUE THREAD

**Issue by SinsOfSeven** [photopea/photopea#7445]
Your Operating System: Win11
Your Web Browser: Brave (Chromium)

We have a large community of users (100k+) who use or make mod games, the images to make game modifications are usually in the dds format, and we've been recommending Paint.net as the best free option. Paint.net has some drawbacks when it comes to editing layers and channels, and recently I was point to Photopea.com as a potential alternative as it had DDS support. However, it seems you do not meet the minimum level of feature support for us to give your site a glowing recommendation to new users, or users who want photoshop-like features without needing to subscribe to it.

I was unable to export DDS files with proper sRGB colorspace, it always exported them as Linear (UNORM vs UNORM_SRGB). The importer properly handles both color spaces, converting them to the colorspace the editor uses. 

Forcing the color profile made a "convert to sRGB" button appear, but this did not change the output on or off, and I didn't really expect it to.
![image](https://github.com/user-attachments/assets/306f8b96-b04d-4cbc-a089-ae878ac0894f)

My work around was to apply a Gamma Correction luminance curve layer, but this is still incorrect, we should just be able to save as linear or sRGB formats for DDS.
![image](https://github.com/user-attachments/assets/36cb6bb3-2cf8-4b83-89c6-8431ed8c092e)

I tried searching for specific implementation details on how to properly store the exported data. But I couldn't really figure out if you need to set the header and store the data with a Gamma correction of 2.2, OR if setting the flag and some gamma value was enough. 
We're very excited to see this addressed, because this web editor is very powerful otherwise. I like that you can save work to your account, I can see a lot of our artists who prefer using their android tablets/ipads trying this out as a nice way to move their work between devices.

References:
[DDS Header](https://learn.microsoft.com/en-us/windows/win32/direct3ddds/dds-header)
[DDS Subheader (pixel format)](https://learn.microsoft.com/en-us/windows/win32/direct3ddds/dds-pixelformat)
[DXGI_FORMAT Enumeration](https://learn.microsoft.com/en-us/windows/win32/api/dxgiformat/ne-dxgiformat-dxgi_format)
[Converting DDS colorspaces](https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/converting-data-color-space)

For my lack of understanding, I've included a portable version of TexConv, perhaps it will be helpful.
[TexConv Portable DLL](https://github.com/matyalatte/Texconv-Custom-DLL)

**Comment by photopea** [comment]
Could you share with me your original file, a DDS which Photopea gives you, and a DDS which you want to get instead?

**Comment by SinsOfSeven** [comment]
I made several examples, just incase using different tools.
[PhotopeaExamples.zip](https://github.com/user-attachments/files/18265681/PhotopeaExamples.zip)

Note, most of the tools refer to UNORM as Linear, and UNORM_SRGB as just sRGB.
All images were exported with alpha, no-mipmaps, and high quality (fine) compression (if given an option).
```
Original BC7_UNORM_SRGB.dds

chaiNNer BC7_UNORM.dds
chaiNNer BC7_UNORM_SRGB.dds
PaintDotNet BC7_UNORM.dds
PaintDotNet BC7_UNORM_SRGB.dds
Photoshop(Intel Plugin) BC7_UNORM.dds
Photoshop(Intel Plugin) BC7_UNORM_SRGB.dds
TexConvDLL BC7_UNORM.dds
TexConvDLL BC7_UNORM_SRGB.dds

Photopea (Manual Gamma) BC7_UNORM.dds
Photopea BC7_UNORM.dds
 ```

The original is pulled directly from directX using something like RenderDoc. Each other example is just a re-export with different tools. 

The Paint.net and TexConvDLL should be very similar, because they use the same DLL, but for the TexConvDLL versions, they were converted to TGA as an intermediate step (inside the blender-dds-plugin). 

I wanted to also include the nVidia photoshop plugin, but the labels are very different so I'm never quite sure if the formats are equivalents, plus it handles the gamma correction as a value. 

I don't know what chaiNNer actually uses for it's DDS support, but I've never had any issues with it's format support.

I read some of the other issues related to DDS, and I don't expect full support of DDS, just the options you have currently are enough, we just need the sRGB option. I did not include any BC1, BC3 examples, but I think it's the same, at the very least it should be easier than bc7, the most involved format.

**Comment by SinsOfSeven** [comment]
I don't know how realistic it is, but I was thinking it might be possible to compile a version of TexConv to wasm, then I was thinking it wouldn't be that hard to also convert the gpu codec to WebGPU and then it stopped being simple, and I'm just excited about WASM and WebGPU, so ignore me unless it's something which is already in your wheelhouse for Photopea.

🤐

**Comment by photopea** [comment]
Could you please describe some steps, what should I do to see a problem that you are talking about?

I opened "chaiNNer BC7_UNORM.dds" and "chaiNNer BC7_UNORM_SRGB.dds" in Photopea and they look the same. Are you saying there is a software which displays them differently? What software it is?

The shorter and clearer your messages are, the sooner I can fix it. Do not send me 11 files if only 1 or 2 files are enough.

**Comment by SinsOfSeven** [comment]
The images look the same because the editing software is interpreting them the same, they are displayed differently when used by directx, and other shading software to render.

This is described in your own documentation. <https://www.photopea.com/learn/color-spaces>

The importing step "knows" how to convert images from whatever colorspace they are in, to a uniform colorspace for editing, before being re-exported into whatever the format expects. In this case, the DDS images are being exported into *only* the linear colorspace, but in realworld usage, they sometimes are expected to be in the sRGB colorspace, depending on what the software expects. The problem is that the image is assumed to be correct, so it just displays the incorrect colors.

### Comparing the example images in use.
The image provided is using the `chaiNNer` textures you've already checked, the top is the `UNORM_SRGB` version, the bottom is just `UNORM`. 
![image](https://github.com/user-attachments/assets/af7f6d37-e7fa-4237-8b8b-794eeb575fc2)

Nearly every image editor at least attempts to convert colorspaces on import so it's not surprising that they appear the same in Photopea, it's actually expected. The issue here is the export format. DirectX does not decide the way the image is interpreted, it just interprets what it is given. 

### What we need from Photopea

I want to be able to select Linear or sRGB on export for the DDS file type and the supported subformats. This is something that almost all image formats have built into them too, including PNG, but most of the time image editors do not expose this part of the file format. In the case of DDS it's an essential part of working with the format, (nearly) every sub format has an sRGB variation. 

As I've said before, I'm unsure of if the process of storing the information is setting a flag in the header, and a conversion value, or if you need to store the colorspace directly into the color of the bits, and the flag just helps the image editors import it correctly.

I just know that there is clearly a difference, and you could use a tool like [DirectXTex TexDiag](https://github.com/Microsoft/DirectXTex/wiki/Texdiag) to see the actual format differences between the otherwise identical looking images.

We really just want to be able to recommend your website as an alternative editor, but this particular feature would be necessary to do so. From a UI/UX perspective, it should be either part of the format dropdowm, or a checkbox on export. Again, we only need sRGB variants of the provided formats (BC1, BC3, BC7, Uncompressed (none)). DDS is a very forgiving format in use, and generally you can make any of these formats work. I also don't expect support for the 64bpp or 128bpp formats, I don't think I saw anything on photopea which does.

### How the other software handles the options.
Paint.net, chaiNNer, Photoshop (Intel Plugin), Blender (Blender-DDS-Plugin). 
![image](https://github.com/user-attachments/assets/0b40610d-fa73-44c9-966a-37350cc9f945)

**Comment by photopea** [comment]
I think I have fixed it, could you try it again? 

It is related only to the BC7 compression, right? The files "chaiNNer BC7_UNORM.dds" and "chaiNNer BC7_UNORM_SRGB.dds" are BC7.

**Comment by SinsOfSeven** [comment]
Brilliant, the site now exported an sRGB variation, but it still needs to be optional. I compared yesterday's export with today's and it had the behavior I expected, and it looks the same when opening it in a photo editor.

Furthermore, it is not exclusive to bc7 compression, ideally it would be available for each compression type, I could provide samples of each format if required. (This time I would only include a single sample of each format).

Photopea | DXGI_FORMAT | Supported
:-|:-|:-
bc1, Linear |BC1_UNORM| ✅
bc1, sRGB |BC1_UNORM_SRGB | ❌ 
bc3, Linear|BC3_UNORM| ✅
bc3, sRGB|BC3__UNORM_SRGB| ❌ 
bc7, Linear|BC7_UNORM| ❌  (after change)
bc7, sRGB|BC7_UNORM_SRGB| ✅ (after change)
none, Linear| R8G8B8A8_UNORM| ✅
none, sRGB|R8G8B8A8_UNORM_SRGB| ❌ 

If you compare Photopea's export options to the Photoshop Intel plugin, you can see the list is almost identical, except for the Linear/sRGB options for each.
![image](https://github.com/user-attachments/assets/ed877404-92c8-44d4-a33c-b7516e7e5039)
![image](https://github.com/user-attachments/assets/1c3c2831-4341-4846-b614-7b4a521ed711)

**Comment by SinsOfSeven** [comment]
I should add, that, only bc7 having both options would be a satisfactory result for the time being.

**Comment by photopea** [comment]
Could you send me examples of DDS files with BC1_UNORM_SRGB and BC3_UNORM_SRGB? I will make Photopea export all files with the SRGB parameter.

**Comment by SinsOfSeven** [comment]
I can send the examples, but I don't recommend making the export type all sRGB or all Linear. For proper DDS support, both options (Linear and sRGB) are more important than supporting many of the extended formats. Having them all export as sRGB will not resolve the issue for us, as it only inverts the issue of them all exporting as linear. The "BC" part is just the block compression algorithm type, and has nothing to do with the colorspace. 

In the [PhotopeaExamples2.zip](https://github.com/user-attachments/files/18270459/PhotopeaExamples2.zip) I've included images which are expected to be in these other colorspaces to try and press the importance. But the given examples are not a universal standard, and it depends on the chosen rendering equation for which format is needed.

TextureType | Expected Colorspace
:-|:-
DiffuseMap | sRGB
Lightmap | Linear
NormalMap | Linear

The DiffuseMap stores color information, so often it is stored in the sRGB colorspace like many other color images, but the Lightmap and NormalMap store normalized data to be used during rendering, where every pixel value is on a linear scale.

The lightmap is a packed texture, where each channel stores important rendering data which, in an sRGB space would destroy the shading results. (But theoretically, could be expected to also be sRGB depending on the rendering equation).
Lightmap Channel | Lightmap Type
:-|:-
Red | Specular
Green | Ambient Occlusion
Blue | Gloss
Alpha | Material ID

I want to stress there is no standard way to know which colorspace for textures will be expected or needed. This has nothing to do with the compression type.

**Comment by photopea** [comment]
This "expected interpretation", like sRGB or linear, is just ONE BYTE in a DDS file. The only thing that Photoepa does differently now, for BC7 export, is this one Byte in the output file. Actual color data stays the same.

I think your software (your games) should allow you to re-define the interpretation of a DDS file, i.e. you can ignore that byte and interpret it however you want. Could you do it somehow, instead of depending on that one Byte?

It is the same as JPG images containing a Byte for the orientation (horizontal / vertical). So when you tell some simple image viewing program to turn your image 90 degrees (permanently), it just changes one Byte of your file, without re-encoding the actual pixel data.

**Comment by SinsOfSeven** [comment]
I understand not wanting to build UI if it's literally a single byte difference, but it does change how different software interprets it. It's not possible to change the software in most cases to interpret the colors differently. It it possible to do it in another piece of software, there are many converters, but we were really hoping to be able to recommend photopea to *lower* the barrier to entry of needing additional software to handle converting between formats. 

If this is something you're not willing to tackle, I understand. It does mean we won't be able to recommend Photopea though. It might only be one byte, is a flag for the GPU to interpret the data differently, using a non-linear function, so while it might seem simple to change, it exists for a very specific reason.

**Comment by photopea** [comment]
Please, stop threatening me with not recommending Photopea to "new users". You already did it for a third time. I dont even know what kind of users you are talking about. You better describe what you want Photopea to do, instead of threatening me.

**Comment by photopea** [comment]
I have added a parameter "DXGI_FORMAT: SRGB". If you enable it, it will export "srgb-ish" DDS files. Does it suit your needs?

**Comment by SinsOfSeven** [comment]
> I have added a parameter "DXGI_FORMAT: SRGB". If you enable it, it will export "srgb-ish" DDS files. Does it suit your needs?

Yes, I was able to validate that every export combination functions as needed, according to the DirectX standard. This is really awesome, since it's hard to find image editing software with good DDS support. Even photoshop needs a plugin.

> Please, stop threatening me with not recommending Photopea to "new users".

My intention was not to threaten you, I just wanted to make it clear this is not "just for me" or "just for a single user's use case".  I truly apologize if this was how it made you feel. We're very glad you were able to add this essential feature support for the DDS format. I would be lying if I said I wasn't trying to incentivize you, but I was worried that it might have been ignored otherwise. Since this was not a problem for me personally, I can easily use these other softwares to deal with DDS files.

> You better describe what you want Photopea to do, instead of threatening me.

I honestly thought I had done a pretty good job laying out what we needed, and you implemented exactly the thing I had asked for. I also provided several examples of how various other software handled it, and tried to provide as much justification as possible. Though I understand it probably came off as pushy.

> ME: From a UI/UX perspective, it should be either part of the format dropdowm, or a checkbox on export.

> I dont even know what kind of users you are talking about.

If you really want to know what kind of users they might be, we wanted to recommend Photopea to people who are new to texture editing for games, through software like ReShade, SpecialK, and 3DMigoto. Generally, any game which uses DirectX as the backend can have textures modded, and the format will always be DDS. We've noticed that many of the younger people joining the community don't have experience with photo-editors, or have trouble with the UIs/Features or Paint.NET and GIMP. We can't recommend GIMP because it doesn't support enough DDS formats, and even though we recommend Paint.NET, it's because it was the ONLY photo editor which supported DDS out of the box, but Paint.NET has it's own issues, like very crappy color channel editing. And while the whole community is over 100k users, the actual people we expect to potentially try photopea is only a few hundred, but those could be people who really like it y'know. I really like it.

I'm happy you were able to implement the requested feature in a short time frame. I wanted to add a feature request tag when I opened the issue, but it didn't let me select one. And lastly, I hope that it was just the way my own messages read which rubbed you the wrong way. I get a little nervous when writing to strangers and sometimes I overdo it on the formal speech.
